docs/JAVA_GRAMMAR_UPDATES.md

```diff
@@ -0,0 +1,225 @@
+```
+
+### Identify New Tokens and Update Lexer Grammar
+
+Not every new language feature requires new tokens. But in the case of the `when` expression,
+we need to introduce a new token to represent the `when` keyword. This requires an update
+to the lexer grammar.
+
+```antlr
+LITERAL_WHEN : 'when' ;
+```
+
+Notes:
+
+- We preface literal tokens with `LITERAL_` to distinguish them from other tokens.
+- Lexer rules are written in uppercase by convention.
+
+We will also need to add a
+new [TokenType](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java)
+in this case. However, we may not need to always add a token type for every new lexer token, it
+depends on the use case.
+
+### Update Parser Grammar
+
+Next, we need to update the parser grammar to recognize the new tokens. We need to add a new parser
+rule
+for the `when` expression. The `when` expression is used in the context of a guard in a switch
+statement or
+switch expression. We need to update the `guard` rule to recognize the `when` expression.
+
+```antlr
+guardedPattern
+    : primaryPattern guard expression
+    ;
+
+// we use the `LAND` token to represent the logical `&&` operator
+guard: ( LAND | LITERAL_WHEN );
+```
+
+Notes:
+
+- Parser rules are written in lower camel case by convention.
+- It is important to consider how many subrules are needed to represent the new language feature.
+  Due to the recursive nature of parsing, Stackoverflow errors can occur if the grammar is too
+  complex with too many subrules.
+- It is good
+  to try to strike a balance between the number of subrules, the complexity of the grammar, and the
+  complexity of JavaAstVisitor to ease maintenance.
+
+### Update JavaAstVisitor
+
+Now, our lexer recognizes the `when` keyword, and our parser recognizes the `when` expression. We
+have updated
+TokenTypes to include `LITERAL_WHEN`. At this point, we are able to parse the new syntax, however,
+the new tokens
+will not appear in the AST unless we update the JavaAstVisitor. This is because our ANTLR grammar
+provides us with a parse tree, which we use the JavaAstVisitor to traverse and build our AST.
+
+```java
+
+@Override
+public DetailAstImpl visitGuardedPattern(JavaLanguageParser.GuardedPatternContext ctx) {
+    // since the `guard` rule is a terminal rule, we need to create a new AST node for it
+    final DetailAstImpl guardAstNode = flattenedTree(ctx.guard());
+    // we add the children of the `primaryPattern` and `expression` rules to the AST node
+    guardAstNode.addChild(visit(ctx.primaryPattern()));
+    guardAstNode.addChild(visit(ctx.expression()));
+    return guardAstNode;
+}
+```
+
+Above is an example of the transformation of the `guardedPattern` rule in the JavaAstVisitor. We
+create a new AST node for the `guard` rule, and add the children of the `primaryPattern` and
+`expression` rules to the AST node. This is a simplified example, and the actual implementation
```

Comment:
What do you think about adding just some note that these 3 (primary, guard, expression) are all children of `guardedPattern` so really, almost all the children should be handled in the visitor.